### ISSUE TITLE
[Bug]: test results are not produced when using custom fixture

### ISSUE THREAD

**Issue by pmzajaczkowski** [microsoft/playwright#7718]
### Playwright version

1.12.3

### Operating system

MacOS, Linux

### What browsers are you seeing the problem on?

Chromium, Firefox, WebKit

### Other information

node: 16.5.0

### What happened? / Describe the bug

Hi, I am seeing the strange behaviour of the debugging tool - it does not produce the results (trace, video, etc.) when a custom fixture is in use.

This is the command definition I use in my package.json file:
`"debug": "PWDEBUG=1 DEBUG=pw:api playwright test"`
I run it as: `npm run debug`

It executes properly - in a meaning of starting the browser and debugger tool and then, after the execution is finished I can see no results produces, especially no trace file.

Here's how my config (playwright.config.ts) looks like:
`import { PlaywrightTestConfig } from "@playwright/test";

const CONFIG_CI: PlaywrightTestConfig = {
    use: {
        // Browser options
        headless: true,

        // Context options
        viewport: { width: 1280, height: 720 },
        ignoreHTTPSErrors: true,

        // Artifacts
        screenshot: "only-on-failure",
        video: "retain-on-failure",
        trace: "on",
    },
    forbidOnly: true, // fail (especially in CI) if there are any tests marked as 'testOnly'
    timeout: 50000,
    retries: 1,
    outputDir: "results",
};

const CONFIG_DEBUG: PlaywrightTestConfig = {
    use: {
        // Browser options
        headless: false,

        // Context options
        viewport: { width: 1280, height: 720 },
        ignoreHTTPSErrors: true,

        // Operation
        slowMo: 2000,

        // Artifacts
        screenshot: "on",
        video: "on",
        trace: "on",
    },
    timeout: 50000,
    retries: 0,
    outputDir: "results",
};

const config = process.env.PWDEBUG && process.env.PWDEBUG == "1" ? CONFIG_DEBUG : CONFIG_CI;
`

Apparently the thing that causes the results not to be produced is a custom fixture - that attempts to perform a login operation only when there is no saved session:
`
import base from '@playwright/test';
import {getBaseURL, getPassword, getUser} from "./config/ConfigProvider";
import {LoginPage} from "../page_objects/LoginPage";

const fs = require('fs');

class LoginInformation {
	user = '';
	pass = '';
	baseURL = '';
}

type TestFixtures = {
	loginInfo: LoginInformation,
	loggedInContext: any,
};

export const test = base.extend<TestFixtures>({
	loginInfo: async ({}, use) => {
		const result = {
			user: getUser(),
			pass: getPassword(),
			baseURL: getBaseURL()
		}
		await use(result);
	},

	loggedInContext: async ({ browser, loginInfo }, use) => {
		if (!fs.existsSync('cookies.json')) {
			console.log("No session yet - logging in...")
			const context = await browser.newContext({ignoreHTTPSErrors: true, storageState: {}})
			const page = await context.newPage()
			const loginPage = new LoginPage(page);
			console.log("Using: ", loginInfo)
			await loginPage.navigate(loginInfo.baseURL);
			await loginPage.fillForm(loginInfo.user, loginInfo.pass);
			await page.waitForLoadState('networkidle');
			await page.screenshot({path: 'screenshots/login/1-form-filled.png'});
			await loginPage.submitForm();
			await page.waitForLoadState('networkidle');
			await page.screenshot({path: 'screenshots/login/1-form-submitted.png'});
			await page.textContent('text=SomeText');
			await page.waitForLoadState('networkidle');
			await page.screenshot({path: 'screenshots/login/2-default-landing-page.png'});

			const cookies = await context.storageState()
			const cookieJson = JSON.stringify(cookies)

			fs.writeFileSync('cookies.json', cookieJson)
			console.log("Login succeeded.")
		} else {
			console.log("Session already established, no need to relogin.")
		}
		const cookies = fs.readFileSync('cookies.json', 'utf8')
		const deserializedCookies = JSON.parse(cookies)
		const context = await browser.newContext({ ignoreHTTPSErrors: true , storageState: deserializedCookies})
		console.log("Using context with established session...")
		await use(context)
	}
});
`
as without this 'feature' the debugging session produced the 'results' directory and trace.zip file is there. On the branch with this fixture the 'results' directory is not even created.

It does not matter if the session has to be established or has been established previously. On the other hand when the test executes the login operations 'by hand' (without the fixture even existing) - the 'results' directory is created correctly.

Maybe I am doing something wrong with this fixture and that's the problem? In such a case I would be grateful for some instructions/advice on how to correctly achieve a result of 'you need to perform the login operation only once' (and it lasts for the time of multiple tests executions).

### Code snippet to reproduce your bug

```shell
import { expect } from '@playwright/test';
import { test } from '../fixtures/BaseFixtures';

test.describe('simple tests', () => {
	test('Sample test 1', async ({loggedInContext, loginInfo}) => {
		const page = await loggedInContext.newPage()
		await page.goto(loginInfo.baseURL)
		const topNavTitle = await page.textContent('text=SomeText')
		expect(topNavTitle).toBe('SomeText')
	})
})
```

### Relevant log output

```shell
$ npm run debug -- -g 'Sample test 1'

> module-name@0.1.0 debug
> PWDEBUG=1 DEBUG=pw:api playwright test "-g" "Sample test 1"

Using config at /.../int-tests/playwright.config.ts

Running 1 test using 1 worker

  ✓ tests/core/mainPage.spec.ts:15:2 › [chromium] simple tests Sample test 1 (12s)
  pw:api => browserType.launch started +0ms
  pw:api <= browserType.launch succeeded +819ms
Session already established, no need to relogin.
  pw:api => browser.newContext started +1ms
  pw:api   navigated to "data:text/html," +919ms
  pw:api   "load" event fired +1ms
  pw:api   "domcontentloaded" event fired +1ms
  pw:api navigating to "https://playwright/index.html", waiting until "load" +9ms
  pw:api   navigated to "https://playwright/index.html" +31ms
  pw:api   "domcontentloaded" event fired +165ms
  pw:api   "load" event fired +18ms
  pw:api <= browser.newContext succeeded +10ms
Using context with established session...
  pw:api => browserContext.newPage started +1ms
  pw:api <= browserContext.newPage succeeded +119ms
  pw:api => page.goto started +0ms
  pw:api   "networkidle" event fired +370ms
  pw:api navigating to "https://...", waiting until "load" +2s
  pw:api   navigated to "https://.../" +1s
  pw:api   "networkidle" event fired +3s
  pw:api   "load" event fired +335ms
  pw:api   "domcontentloaded" event fired +1ms
  pw:api <= page.goto succeeded +2s
  pw:api => page.textContent started +0ms
  pw:api   retrieving textContent from "text=SomeText" +1ms
  pw:api   navigated to "https://..." +1s
  pw:api   navigated to "https://..." +1ms
  pw:api   selector resolved to <h3 class="sc-coaFxK ffdDgP">SomeText</h3> +7ms
  pw:api <= page.textContent succeeded +95ms
  pw:api => browser.close started +11ms
  pw:api <= browser.close succeeded +120ms

  1 passed (13s)
```

**Comment by aslushnikov** [comment]
Over to Joel who saw this bug already.

Speculative suggestion: introduce a `browser.setDefaultContextOptions()` method to support this usecase.

**Comment by sammyhass** [comment]
Any updates to this? Have been having a similar issue for a few days now - using `test.describe` to write my tests

------
playwright.config.ts
`const config: PlaywrightTestConfig = {
  retries: 2,
  preserveOutput: 'always',
  workers: 1,
  use: {
    headless: true,
    slowMo: 50,
    viewport: { width: 1440, height: 900 },
    contextOptions: {
      ignoreHTTPSErrors: true,
      permissions: ['geolocation'],
      geolocation: { latitude: 51, longitude: 50 },
    },
    screenshot: 'only-on-failure',
    video: 'on',
    trace: 'on',
  },
};`

------
However, this still will not take screenshots when tests fail.

**Comment by pmzajaczkowski** [comment]
I am not sure, but I suppose that this whole problem might be caused by instantiating custom contexts within the testing code. It seems that any configuration options provided there cause the config file options to be totally ignored.

I mean operations like these:
`const context = await browser.newContext({ignoreHTTPSErrors: true, storageState: {}})`

**Comment by sammyhass** [comment]
> I am not sure, but I suppose that this whole problem might be caused by instantiating custom contexts within the testing code. It seems that any configuration options provided there cause the config file options to be totally ignored.
> 
> I mean operations like these:
> `const context = await browser.newContext({ignoreHTTPSErrors: true, storageState: {}})`

Perfect this sorted it straight away, thank you!

**Comment by dgozman** [comment]
Fixed by #8083.
